Documentation for data directory. (#1125)
algorand/indexer: docs/DataDirectory.md

# Indexer Data Directory

The Indexer data directory is the location where the Indexer can store and/or load data needed for runtime operation and configuration. It is a required argument for Indexer daemon operation. Supply it to the Indexer via the `--data-dir` flag.

# Storage Requirements

As of mid-2022, approximately 15GiB for Mainnet.

# Configuration Files

The data directory is the first place to check for different configuration files, for example:
- `indexer.yml` - Indexer Configuration File
- `api_config.yml` - API Parameter Enable/Disable Configuration File

> # Account Cache

Author: I'll list the files it creates.

Indexer writers keep an account cache in the data directory. This cache is used during block processing to compute things like the new account balances after processing transactions. Prior to this local cache, the database was queried on each round to fetch the initial account states.

> ## Read-Only Mode

Reviewer: If an indexer is started in writer mode and then is re-started in reader mode, will these files be there or will they be deleted by the indexer?

Author: They wouldn't be deleted, but wouldn't be updated either.

The account cache is not required when in read-only mode. While the data directory is still required, it will only be used for configuration.

# Initialization

If a new data directory must be created, the following process should be used:
1. Review the Indexer log to find the most recent round that was processed. For example, `22212765` in the following line:
   ```
   {"level":"info","msg":"round r=22212765 (49 txn) imported in 139.782694ms","time":"2022-07-18T19:23:13Z"} 
   ```
2. Lookup the most recent catchpoint **without going over** from the list for your network:
   [Mainnet]()
   [Testnet]()
   [Betanet]()

> 3. Supply the catchpoint label when starting Indexer using the command line setting `--catchpoint 6500000#1234567890ABCDEF01234567890ABCDEF0`, setting `catchpoint` in `indexer.yml`, or setting the `INDEXER_CATCHPOINT` environment variable.

Reviewer 2: Is it possible for the indexer to detect the most recent round that was processed, and then automatically download the needed catchpoint?

Author: Good idea, yes it would totally be possible to do that. I'll create an issue to track that.

Author: Created: https://github.com/algorand/indexer/issues/1132

- `indexer.yml` - Indexer Configuration File
- `api_config.yml` - API Parameter Enable/Disable Configuration File

# Account Cache

Indexer writers maintain an account cache in the data directory. This cache is used during block processing to compute things like the new account balances after processing transactions. Prior to this local cache, the database was queried on each round to fetch the initial account states.

The following files are created:
- ledger.block.sqlite
- ledger.block.sqlite-shm
- ledger.block.sqlite-wal
- ledger.tracker.sqlite
- ledger.tracker.sqlite-shm
- ledger.tracker.sqlite-wal

## Read-Only Mode

The account cache is not required when in read-only mode. While the data directory is still required, it will only be used for configuration.

# Initialization

If a new data directory must be created, the following process should be used:
1. Review the Indexer log to find the most recent round that was processed. For example, `22212765` in the following line:
   ```
   {"level":"info","msg":"round r=22212765 (49 txn) imported in 139.782694ms","time":"2022-07-18T19:23:13Z"} 
   ```

> 2. Lookup the most recent catchpoint **without going over** from the list for your network. For example, with `22212765` from step 1, on mainnet you would choose `22210000#MZZIOYXYPPGNYRQHROXCPILIWIMQQRN7ZNLQJVM2QVSKT3QX6O4A`. A full list of available catchpoints for each network can be found at the following links:

Reviewer 3: ```suggestion
2. Lookup the most recent catchpoint for your network **without going over the indexer's current round** from the following links. For example, with `22212765` from step 1, on mainnet you would choose `22210000#MZZIOYXYPPGNYRQHROXCPILIWIMQQRN7ZNLQJVM2QVSKT3QX6O4A` from the Mainnet consolidated catchpoint list. 

Full list of available catchpoints for each network can be found at the following links:
```

Full list of available catchpoints for each network can be found at the following links:
   - [Mainnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/mainnet_catchpoints.txt)
   - [Testnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/testnet_catchpoints.txt)
   - [Betanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/betanet_catchpoints.txt)

>    - [Devnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/devnet_catchpoints.txt)

Reviewer 3: ```suggestion
```
Remove DevNet too since the public shouldn't use DevNet. We can document the other catchpoint lists internally.

# Initialization

If a new data directory must be created, the following process should be used:
1. Review the Indexer log to find the most recent round that was processed. For example, `22212765` in the following line:
   ```
   {"level":"info","msg":"round r=22212765 (49 txn) imported in 139.782694ms","time":"2022-07-18T19:23:13Z"} 
   ```
2. Lookup the most recent catchpoint for your network **without going over the indexer's current round** from the following links. For example, with `22212765` from step 1, on mainnet you would choose `22210000#MZZIOYXYPPGNYRQHROXCPILIWIMQQRN7ZNLQJVM2QVSKT3QX6O4A` from the Mainnet consolidated catchpoint list. 
Full list of available catchpoints for each network can be found at the following links:
   - [Mainnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/mainnet_catchpoints.txt)
   - [Testnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/testnet_catchpoints.txt)
   - [Betanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/betanet_catchpoints.txt)

>    - [Alphanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/alphanet_catchpoints.txt)

Author: @chaihoang suggested removing this.
```suggestion
```

Reviewer 3: We should probably remove devnet too then right?

>    [Mainnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/mainnet_catchpoints.txt)
>    [Testnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/testnet_catchpoints.txt)
>    [Betanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/betanet_catchpoints.txt)
>    [Alphanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/alphanet_catchpoints.txt)
>    [Devnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/devnet_catchpoints.txt)

Author: ```suggestion
   - [Mainnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/mainnet_catchpoints.txt)
   - [Testnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/testnet_catchpoints.txt)
   - [Betanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/betanet_catchpoints.txt)
   - [Alphanet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/alphanet_catchpoints.txt)
   - [Devnet](https://algorand-catchpoints.s3.us-east-2.amazonaws.com/consolidated/devnet_catchpoints.txt)
```

- `indexer.yml` - Indexer Configuration File
- `api_config.yml` - API Parameter Enable/Disable Configuration File

# Account Cache

Indexer writers maintain an account cache in the data directory. This cache is used during block processing to compute things like the new account balances after processing transactions. Prior to this local cache, the database was queried on each round to fetch the initial account states.

The following files are created:
- ledger.block.sqlite
- ledger.block.sqlite-shm
- ledger.block.sqlite-wal
- ledger.tracker.sqlite
- ledger.tracker.sqlite-shm
- ledger.tracker.sqlite-wal

## Read-Only Mode

The account cache is not required when in read-only mode. While the data directory is still required, it will only be used for configuration.

# Initialization

If a new data directory must be created, the following process should be used:
1. Review the Indexer log to find the most recent round that was processed. For example, `22212765` in the following line:
   ```
   {"level":"info","msg":"round r=22212765 (49 txn) imported in 139.782694ms","time":"2022-07-18T19:23:13Z"} 
   ```

> 2. Lookup the most recent catchpoint for your network **without going over the indexer's current round** from the following links. For example, with `22212765` from step 1, on mainnet you would choose `22210000#MZZIOYXYPPGNYRQHROXCPILIWIMQQRN7ZNLQJVM2QVSKT3QX6O4A` from the Mainnet consolidated catchpoint list. 
> 
> Full list of available catchpoints for each network can be found at the following links:

Author: This extra newline broke the markdown, trying with one newline.
```suggestion
2. Lookup the most recent catchpoint for your network **without going over the indexer's current round** from the following links. For example, with `22212765` from step 1, on mainnet you would choose `22210000#MZZIOYXYPPGNYRQHROXCPILIWIMQQRN7ZNLQJVM2QVSKT3QX6O4A` from the Mainnet consolidated catchpoint list. 
Full list of available catchpoints for each network can be found at the following links:
```
